x-pack/legacy/plugins/ml/public/util/custom_url_utils.ts

```diff
@@ -4,37 +4,42 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
-
-
 // utility functions for handling custom URLs
 
 import _ from 'lodash';
 import moment from 'moment';
 
 import { parseInterval } from '../../common/util/parse_interval';
-import {
-  escapeForElasticsearchQuery,
-  replaceStringTokens } from './string_utils';
+import { escapeForElasticsearchQuery, replaceStringTokens } from './string_utils';
+import { UrlConfig, KibanaUrlConfig } from '../../common/types/custom_urls';
+import { AnomalyRecordSource, AnomalyRecordDoc } from '../../common/types/anomalies';
 
 // Value of custom_url time_range property indicating drilldown time range is calculated automatically
 // depending on the context in which the URL is being opened.
 const TIME_RANGE_AUTO = 'auto';
 
 // Replaces the $ delimited tokens in the url_value of the custom URL configuration
 // with values from the supplied document.
-export function replaceTokensInUrlValue(customUrlConfig, jobBucketSpanSecs, doc, timeFieldName) {
+export function replaceTokensInUrlValue(
+  customUrlConfig: UrlConfig | KibanaUrlConfig,
+  jobBucketSpanSecs: number,
+  doc: AnomalyRecordDoc,
+  timeFieldName: 'timestamp' | string
+) {
   // If urlValue contains $earliest$ and $latest$ tokens, add in times to the test doc.
   const urlValue = customUrlConfig.url_value;
   const timestamp = doc[timeFieldName];
-  const timeRangeInterval = parseInterval(customUrlConfig.time_range);
+  const timeRangeInterval =
+    'time_range' in customUrlConfig ? parseInterval(customUrlConfig.time_range) : null;
+  const record = { ...doc } as AnomalyRecordSource;
```

Comment:
Nit: Just wondering would `const record: AnomalyRecordSource = { ...doc };` work here instead of `as ...` or does it clash with the use of the spread? Or is it because `earliest/latest` is only added later on?

Reply:
`const record: AnomalyRecordSource = { ...doc };` isn't going to work. Yes TS requires `earliest` and `latest` to be defined during declaration.

Reply:
Ok got it! To solve this you could refactor to something like:

```js
const record: AnomalyRecordSource = {
  ...doc,
  earliest: getEarliest(doc),
  latest: getLatest(doc),
};
```

Looking at the current code, are there cases when `earliest/latest` could stay empty? It looks like the code doesn't add anything if the lines `if (urlValue.includes('$earliest/lastest$')) {` are `false`. If that's the case maybe the interface needs `?` for `earliest/latest`?